File: MirageMenuBar/Sources/MirageMenuBar/GaiaProtocol.swift
import Foundation

enum MirageError: Error, LocalizedError {
    case message(String)

    var errorDescription: String? {
        switch self {
        case .message(let message): return message
        }
    }
}

struct GaiaPDU {
    let vendor: Int
    let feature: Int
    let type: Int
    let command: Int
    let payload: [UInt8]
}

enum AncMode: UInt8, CaseIterable {
    case off = 0
    case adaptive = 1
    case transparency = 2
    case wind = 3
    case anc = 4

    var title: String {
        switch self {
        case .off: return "关闭"
        case .adaptive: return "自适应"
        case .transparency: return "通透"
        case .wind: return "抗风噪"
        case .anc: return "降噪"
        }
    }
}

struct BatteryLevels {
    var left: UInt8?
    var right: UInt8?
    var chargingCase: UInt8?

    static func parse(_ payload: [UInt8]) -> BatteryLevels {
        var values: [UInt8: UInt8] = [:]
        var index = 0
        while index + 1 < payload.count {
            let value = payload[index + 1]
            values[payload[index]] = value <= 100 ? value : nil
            index += 2
        }
        return BatteryLevels(left: values[1], right: values[2], chargingCase: values[3])
    }
}

struct MultipointDevice: Hashable {
    let slot: String
    let status: UInt8
    let address: String
    let addressBytes: [UInt8]
    let name: String
}

enum GaiaCodec {
    static let vendor = 0x001D

    static func makePDU(feature: Int, command: Int, payload: [UInt8] = []) -> [UInt8] {
        let commandValue = (feature << 9) | (command & 0x7F)
        return [
            UInt8((vendor >> 8) & 0xFF), UInt8(vendor & 0xFF),
            UInt8((commandValue >> 8) & 0xFF), UInt8(commandValue & 0xFF),
        ] + payload
    }

    static func wrapV4(_ pdu: [UInt8]) -> [UInt8] {
        let payloadLength = max(0, pdu.count - 4)
        if payloadLength <= 0xFF {
            return [0xFF, 0x04, 0x00, UInt8(payloadLength)] + pdu
        }
        return [
            0xFF, 0x04, 0x02,
            UInt8((payloadLength >> 8) & 0xFF), UInt8(payloadLength & 0xFF),
        ] + pdu
    }

    static func parsePDU(_ bytes: [UInt8]) -> GaiaPDU? {
        guard bytes.count >= 4 else { return nil }
        let vendor = (Int(bytes[0]) << 8) | Int(bytes[1])
        let value = (Int(bytes[2]) << 8) | Int(bytes[3])
        return GaiaPDU(
            vendor: vendor,
            feature: (value >> 9) & 0x7F,
            type: (value >> 7) & 0x03,
            command: value & 0x7F,
            payload: Array(bytes.dropFirst(4))
        )
    }
}

File: MirageController/app/src/main/java/com/cgluwxh/miragecontroller/GaiaProtocol.kt
package com.cgluwxh.miragecontroller

data class GaiaPdu(val vendor: Int, val feature: Int, val type: Int, val command: Int, val payload: ByteArray)

enum class AncMode(val raw: Int, val title: String) {
    OFF(0, "关闭"), ADAPTIVE(1, "自适应"), TRANSPARENCY(2, "通透"), WIND(3, "抗风噪"), ANC(4, "降噪");
    companion object { fun fromRaw(value: Int?) = entries.firstOrNull { it.raw == value } }
}

fun nextNoiseMode(current: AncMode?): AncMode = when (current) {
    AncMode.ANC -> AncMode.ADAPTIVE
    AncMode.ADAPTIVE -> AncMode.WIND
    AncMode.WIND -> AncMode.ANC
    else -> AncMode.ANC
}

data class BatteryLevels(val left: Int? = null, val right: Int? = null, val chargingCase: Int? = null) {
    companion object {
        fun parse(payload: ByteArray): BatteryLevels {
            val values = mutableMapOf<Int, Int?>()
            var index = 0
            while (index + 1 < payload.size) {
                val value = payload[index + 1].toInt() and 0xff
                values[payload[index].toInt() and 0xff] = value.takeIf { it <= 100 }
                index += 2
            }
            return BatteryLevels(values[1], values[2], values[3])
        }
    }
}

data class MultipointDevice(val slot: String, val status: Int, val address: String, val addressBytes: ByteArray, val name: String)
data class DeviceSnapshot(val battery: BatteryLevels, val anc: AncMode?, val multipointEnabled: Boolean, val timeout: Int, val devices: List<MultipointDevice>)

object GaiaCodec {
    const val VENDOR = 0x001d

    fun frame(feature: Int, command: Int, payload: ByteArray = byteArrayOf()): ByteArray {
        val commandValue = (feature shl 9) or (command and 0x7f)
        val pdu = byteArrayOf((VENDOR ushr 8).toByte(), VENDOR.toByte(), (commandValue ushr 8).toByte(), commandValue.toByte()) + payload
        return byteArrayOf(0xff.toByte(), 0x04, 0x00, (pdu.size - 4).toByte()) + pdu
    }

    fun parse(bytes: ByteArray): GaiaPdu? {
        if (bytes.size < 4) return null
        val vendor = ((bytes[0].toInt() and 0xff) shl 8) or (bytes[1].toInt() and 0xff)
        val value = ((bytes[2].toInt() and 0xff) shl 8) or (bytes[3].toInt() and 0xff)
        return GaiaPdu(vendor, (value ushr 9) and 0x7f, (value ushr 7) and 3, value and 0x7f, bytes.copyOfRange(4, bytes.size))
    }
}
